[quantization] Use ConvPackedParams everywhere

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30199 [quantization] Use ConvPackedParams everywhere**

This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

Commit:
[quantization] Use ConvPackedParams everywhere

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -317,27 +339,24 @@ def _get_name(self):
 
     def _weight_bias(self):
-        return torch.ops.quantized.conv3d_unpack(self._packed_params)
+        return self._packed_params._weight_bias()
```
this is conv2d_unpack....we'll need to have Conv3dPackedParams module for this..

jamesr66a replied:
Are there any tests or checks for this? It all passes just fine in the test suite....

jerryzh168 replied:
cc @dskhudia @BIT-silence do we have tests for conv3d_prepack and conv3d_unpack?

jamesr66a replied:
@jerryzh168 figured out the missing test coverage, please see https://github.com/pytorch/pytorch/pull/30247

dskhudia replied:
> cc @dskhudia @BIT-silence do we have tests for conv3d_prepack and conv3d_unpack?

Yes. There is.

jerryzh168 commented:
btw LinearPackedParams and ConvPackedParams are temporary workaround in graphmode to support serialization of packed values. And long term support would be TorchBind. So what about eager mode? do we always need to do this workaround?

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

jamesr66a commented:
@jerryzh168 this is the correct way to do it, I've already done the same refactoring for RNNs. We can upgrade to using torchbind in the future, but after this change that should just be a 1:1 replacement of the TorchScript *PackedParams modules with the torchbind class

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -16,10 +16,103 @@
 
+
+class ConvPackedParams(torch.nn.Module):
+    def __init__(self, spatial_dim):
```
maybe make default value 2?

jamesr66a replied:
I don't think a default value saves us much here, and it is bound to cause trouble when others are trying to use this class

jerryzh168 replied:
I see, makes sense

jerryzh168 reviewed (approved):
LGTM

raghuramank100 commented:
Can you also add versioning support like was done in https://github.com/pytorch/pytorch/pull/30481 as part of this PR?

raghuramank100 reviewed (dismissed):
Please include support for older versions of the state dict as part of this PR so that torchvision support for quantized models is not broken (just like https://github.com/pytorch/pytorch/pull/30481)

Commit:
Update on "[quantization] Use ConvPackedParams everywhere"


This makes things simpler and is probably how we should've done it in the first place

Differential Revision: [D18628004](https://our.internmc.facebook.com/intern/diff/D18628004)

[ghstack-poisoned]
